feat: windows app icon

Klemen2 opened the pull request:
![image](https://github.com/user-attachments/assets/07698d79-ad7f-45b9-84e0-3e4b81ef8f6b)

Closes: #3648
Implement #3596 for windows

Windows app resources will only be build with --release flag, otherwise it will be embedded (no app icon (thumbnail) or metadata when installed)

This pr also reduces the binary size since including icon is now deduplicated.

Added functions: default_icon, icon_from_memory, icon_from_path

Tested only on windows 11

photino commented:
Why didn't previous versions support Windows app icons?

rohitsangwan01 commented on packages/cli/src/build/request.rs:
```diff
@@ -4575,4 +4580,110 @@ __wbg_init({{module_or_path: "/{}/{wasm_path}"}}).then((wasm) => {{
     }
+
+    fn winres_icon_path(&self) -> crate::winres::WindowsResourceLinker {
+        todo!()
```
This is causing issue for me, any idea when this can be merged ?

Klemen2 replied:
The [main](https://github.com/Klemen2/dioxus/tree/main) fork is synced with main and has this PR merged in it @rohitsangwan01

Klemen2 commented:
> Anything left i can help with to bring this over the finish line? 0.7 is is nearing its release and i would love to not use a workaround for my current releases with 0.6 to have app icons.

There are a few changes that needs to be done. 
1. testing bundling apps for windows on mac/linux
2. It should probably not fail to build for windows if you do not have the tooling, it should only fail for bundling and release builds this requires changing how default icon is handled on windows (embedded with include_bytes instead of loaded from_resource) and if that's the case notify the user that there will be no icon for the app
3. I noticed that the logs "Compiling Windows resource file with ..." gets outputted twice, it should just be once
4. default icon handling for other platforms (related to 2. but ideally it would get it from the app itself and fallback  to include_bytes for non release/bundle) -- not required for this PR, but would be nice for consistency
5. caching the windows resource file so it only compiles when the associated files changes  -- not required for this PR

I think this is everything unless @jkelleyrtp has anything to add

pythoneer commented:
I think the work here is relevant https://github.com/DioxusLabs/dioxus/pull/4958 idk.

Klemen2 commented:
winres was allready doing compilation of the icon in the cli and was not using any cargo / build script - basically what cloud suggested, I removed a bunch of comments and one missed line that still had a cargo in winres and also changed the structure a little

I haven't tested the gnu toolkit or on other platforms, only mscv on windows only

`println!("cargo:rustc-env=DIOXUS_APP_ICON=./assets/default_icon.png")` was done so that #3596 could be closed - now it's only implemented for windows

pythoneer commented:
relevant PR https://github.com/DioxusLabs/dioxus/pull/5328

pythoneer commented:
I think https://github.com/DioxusLabs/dioxus/pull/5412 and https://github.com/DioxusLabs/dioxus/pull/5328 caused conflicts again. Not sure how this affects this PR :(

jkelleyrtp commented:
Would love to merge this. What's the current state?

After digging into windows bundling recently, I have several thoughts:
- `dx bundle` shells out to wix / nsi which have options to include icons. For many users this could just be enough? You don't get icons during dev, but you do in bundle, which is really what matters.
- There is a crate called sui which lets you inject icons and files into binaries after they've been built (https://github.com/denoland/sui). This might be pretty simple.
- We can just add the icon via the linker - ie dx runs `cargo rustc -- -C link-arg=app.res` automatically. I think that's what's happening in this PR?

It seems that on many platforms we need some sort of pre-processing step for icon generation (android as well, and mac/ios). This adds extra overhead in dev (slower dev-server starts until we properly parallelize things), which is why we've deferred it to bundle. 

I'm happy to add this step in dev too if folks think it's important.

Klemen2 commented on packages/desktop/src/default_icon.rs:
```diff
@@ -0,0 +1,173 @@
+
+// preferably this would have platform specific implementations, not just for windows
+#[cfg(not(target_os = "windows"))]
+static DEFAULT_ICON: &[u8] = include_bytes!("./assets/default_icon.png");
```
as for other platforms than windows, the easy way would be to replace this with an user icon for the default icon from toml to work, so if thats fine I can do that

Klemen2 replied:
there is probably a better way of doing that than this, but I don't think that I have time for that

Klemen2 replied:
as far as I can tell, the window icon matters only on windows, wayland and x11, for others it shouldn't have any effects for the current stable tao/winit

for tray icon it matters on all desktop platforms, so it probably would be a good idea to implement it for all desktop platforms, not just windows

I've changed it right now so it just returns an error which makes it so there is no icon

Klemen2 reviewed (commented):
As far as I know, wix / nsi settings only add icon to the installer, not the actual app.
the pr is basically doing cargo rustc -- -C link-arg=app.res where res needs to be compiled with rc.exe (or equivalent) because windows is special.

I don't think that dev builds necessarily need an icon, so atm you just get a warning if it can't be compiled.

I think that I only need to update a couple of things which I can probably do this weekend.

Klemen2 committed:
default icon

Klemen2 committed:
cli

Klemen2 committed:
fix typos, fmt + icon fix

I don't think it's a good idea to do
[default.extend-words]
fo = "fo"

 as it's a very common mispelling for 'for'

Klemen2 committed:
fix clippy

Klemen2 committed:
return error on non implemented platforms instead of default dx icon

Klemen2 committed:
fix

Klemen2 committed:
Merge branch 'main' into windows-app-icon

Klemen2 committed:
fix

fix the amazing merge
renamed winres to windows
moved windows winres build request to windows

Klemen2 committed:
fix typos

Klemen2 committed:
fix schema

Klemen2 committed:
Update schema.json

Klemen2 committed:
fix schema again

Klemen2 commented:
I'd say this is ready, unless you want any changes in the windows scope.

jkelleyrtp committed:
use default icon if no fallback

jkelleyrtp committed:
wire up the rest of windows

jkelleyrtp committed:
fmt

jkelleyrtp commented:
Thanks so much!

adamgreg commented:
The addition of `ApplicationConfig::name` in this PR is very interesting! It looks like it's not really used yet, but is it intended as a way to override the application name in bundles, so it doesn't have to be a PascalCase version of the crate name? That would be really useful for my app.

Is anyone working on that? I couldn't find an open PR. I'd be willing to give it a try myself if that would be welcome.

Klemen2 commented:
right now it's only used for the windows product name (app -> right click -> properties -> details)
the app name is not yet affected but fell free to implement it

adamgreg commented:
> right now it's only used for the windows product name (app -> right click -> properties -> details) the app name is not yet affected but fell free to implement it

FYI I just raised #5533 for this
